torchvision/prototype/datapoints/_dataset_wrapper.py

```diff
@@ -0,0 +1,250 @@
+# type: ignore
+
+from __future__ import annotations
+
+import contextlib
+
+import functools
+from collections import defaultdict
+from typing import Any, Callable, Dict, List, Optional, Tuple
+
+import PIL.Image
+import torch
+from torch.utils.data import Dataset
+from torchvision import datasets
+from torchvision.prototype import datapoints
+from torchvision.prototype.transforms import functional as F
+
+__all__ = ["wrap_dataset_for_transforms_v2"]
+
+_WRAPPERS = {}
+
+
+# TODO: naming!
+def wrap_dataset_for_transforms_v2(dataset: datasets.VisionDataset) -> _VisionDatasetDatapointWrapper:
+    wrapper = _WRAPPERS.get(type(dataset))
+    if wrapper is None:
+        raise TypeError
+    return _VisionDatasetDatapointWrapper(dataset, wrapper)
+
+
+class _VisionDatasetDatapointWrapper(Dataset):
+    def __init__(self, dataset: datasets.VisionDataset, wrapper) -> None:
+        self.vision_dataset = dataset
+        self.wrapper = wrapper
+
+        # We need to disable the transforms on the dataset here to be able to inject the wrapping before we apply the
+        # transforms
+        self.transform, dataset.transform = dataset.transform, None
+        self.target_transform, dataset.target_transform = dataset.target_transform, None
+        self.transforms, dataset.transforms = dataset.transforms, None
+
+    def __getattr__(self, item: str) -> Any:
+        with contextlib.suppress(AttributeError):
+            return object.__getattribute__(self, item)
+
+        return getattr(self.vision_dataset, item)
+
+    def __getitem__(self, idx: int) -> Any:
+        # This gets us the raw sample since we disabled the transforms for the underlying dataset in the constructor
+        # of this class
+        sample = self.vision_dataset[idx]
+
+        sample = self.wrapper(self.vision_dataset, sample)
+
+        # We don't need to care about `transform` and `target_transform` here since `VisionDataset` joins them into a
+        # `transforms` internally:
```

Comment:
So do we actually need these 2 lines above in `__init__`? 

```py
        self.transform, dataset.transform = dataset.transform, None
        self.target_transform, dataset.target_transform = dataset.target_transform, None
```

Looks like they'd still be available through our custom `__getattr__`?

Reply:
Fair point. I guess we only need to disable `transforms`.

Reply:
I've reviewed this again and it was correct to disable everything. At this stage we only need to care about `transforms` since it is guaranteed that it includes all functionality even if the user passed `transform` and `target_transform` individually. However, most datasets don't use `transforms` in `__getitem__`, but rather the individual ones. Since our goal is to get the untransformed sample, we need to disable everything that might change it before we get our hands on it.

I've expanded the comment to explain this better.